File: Day #1529/Solution.java
// Prune binary tree: remove subtrees containing only 0s (no 1 descendant).
// Post-order recursion. O(n) time, O(h) recursion stack.
import java.util.*;

public class Solution {
    static class Node {
        int val;
        Node left, right;
        Node(int v) { val = v; }
    }

    static Node prune(Node root) {
        if (root == null) return null;
        root.left = prune(root.left);
        root.right = prune(root.right);
        if (root.val == 0 && root.left == null && root.right == null) return null;
        return root;
    }

    static void printLevelOrder(Node root) {
        Queue<Node> q = new LinkedList<>();
        q.add(root);
        List<String> out = new ArrayList<>();
        while (!q.isEmpty()) {
            Node n = q.poll();
            if (n == null) { out.add("null"); continue; }
            out.add(String.valueOf(n.val));
            q.add(n.left);
            q.add(n.right);
        }
        while (!out.isEmpty() && out.get(out.size() - 1).equals("null")) out.remove(out.size() - 1);
        System.out.println("[" + String.join(", ", out) + "]");
    }

    public static void main(String[] args) {
        Node root = new Node(0);
        root.left = new Node(1);
        root.right = new Node(0);
        root.right.left = new Node(1);
        root.right.right = new Node(0);
        root.right.left.left = new Node(0);
        root.right.left.right = new Node(0);

        root = prune(root);
        printLevelOrder(root); // [0, 1, 0, null, null, 1]
    }
}

File: Day #1529/solution.go
// Prune binary tree: remove subtrees containing only 0s (no 1 descendant).
// Post-order recursion. O(n) time, O(h) recursion stack.
package main

import (
	"fmt"
	"strconv"
	"strings"
)

type Node struct {
	Val         int
	Left, Right *Node
}

func prune(root *Node) *Node {
	if root == nil {
		return nil
	}
	root.Left = prune(root.Left)
	root.Right = prune(root.Right)
	if root.Val == 0 && root.Left == nil && root.Right == nil {
		return nil
	}
	return root
}

func levelOrder(root *Node) string {
	queue := []*Node{root}
	var out []string
	for len(queue) > 0 {
		n := queue[0]
		queue = queue[1:]
		if n == nil {
			out = append(out, "null")
			continue
		}
		out = append(out, strconv.Itoa(n.Val))
		queue = append(queue, n.Left, n.Right)
	}
	for len(out) > 0 && out[len(out)-1] == "null" {
		out = out[:len(out)-1]
	}
	return "[" + strings.Join(out, ", ") + "]"
}

func main() {
	root := &Node{Val: 0,
		Left: &Node{Val: 1},
		Right: &Node{Val: 0,
			Left:  &Node{Val: 1, Left: &Node{Val: 0}, Right: &Node{Val: 0}},
			Right: &Node{Val: 0}}}
	fmt.Println(levelOrder(prune(root))) // [0, 1, 0, null, null, 1]
}
